### PR TITLE
bpo-40077: Convert _jsonmodule to use PyType_FromSpec.

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue40077

### CODE REVIEW

File: Modules/_json.c [hunk 1]
```diff
@@ -628,40 +634,37 @@ py_encode_basestring(PyObject* Py_UNUSED(self), PyObject *pystr)
...
 static void
-scanner_dealloc(PyObject *self)
+scanner_dealloc(PyScannerObject *self)
 {
+    PyTypeObject *tp = Py_TYPE(self);
     /* bpo-31095: UnTrack is needed before calling any callbacks */
     PyObject_GC_UnTrack(self);
     scanner_clear(self);
-    Py_TYPE(self)->tp_free(self);
+    freefunc free_func = PyType_GetSlot(tp, Py_tp_free);
```
Comment (vstinner) [inline Modules/_json.c]: Even if the type is allocated on the heap, you can still use `Py_TYPE(self)->tp_free(self);`.

Same remark for similar changes below.

File: Modules/_json.c [hunk 2]
```diff
@@ -628,40 +634,37 @@ py_encode_basestring(PyObject* Py_UNUSED(self), PyObject *pystr)
 }
 
 static void
-scanner_dealloc(PyObject *self)
+scanner_dealloc(PyScannerObject *self)
```
Comment (vstinner) [inline Modules/_json.c]: I'm not sure why you replace `PyObject *self` with `PyScannerObject *self`. No PyScannerObject field is accessed here.

File: Modules/_json.c [hunk 3]
```diff
@@ -628,40 +634,37 @@ py_encode_basestring(PyObject* Py_UNUSED(self), PyObject *pystr)
...
+    free_func(self);
+    Py_DECREF(tp);
 }
 
 static int
-scanner_traverse(PyObject *self, visitproc visit, void *arg)
+scanner_traverse(PyScannerObject *self, visitproc visit, void *arg)
 {
-    PyScannerObject *s;
-    assert(PyScanner_Check(self));
```
Comment (vstinner) [inline Modules/_json.c:643]: I would prefer to keep the assertion for safety. Same comment for other removed assertions.

See: https://en.wikipedia.org/wiki/Wikipedia:Chesterton's_fence

Don't remove an assertion if you don't know why it was added ;-)

File: Modules/_json.c [hunk 4]
```diff
@@ -1209,47 +1210,24 @@ scanner_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
-    0,                    /* tp_init */
-    0,/* PyType_GenericAlloc, */        /* tp_alloc */
-    scanner_new,          /* tp_new */
-    0,/* PyObject_GC_Del, */              /* tp_free */
+PyType_Spec PyScannerType_spec = {
+    "_json.Scanner",
+    sizeof(PyScannerObject),
+    0,
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    PyScannerType_slots,
```
Comment (vstinner) [inline Modules/_json.c]: While you're modifying the code... I would prefer to use C99 `.field = value,` syntax, like `.tp_name = "_json.Scanner",`.

File: Modules/_json.c [hunk 5]
```diff
@@ -1866,25 +1814,59 @@ PyDoc_STRVAR(module_doc,
...
 {
-    if (PyType_Ready(&PyScannerType) < 0) {
+    PyObject *PyScannerType = PyType_FromSpec(&PyScannerType_spec);
+    if (PyScannerType == NULL) {
         return -1;
     }
-    if (PyType_Ready(&PyEncoderType) < 0) {
+    get_json_state(module)->PyScannerType = PyScannerType;
+    Py_INCREF(PyScannerType);
+    if (PyModule_AddObject(module, "make_scanner", get_json_state(module)->PyScannerType) < 0) {
```
Comment (vstinner) [inline Modules/_json.c]: `Py_DECREF((PyObject*)&PyScannerType);` is still needed here, no? Same remark for the similar code below.

File: Modules/_json.c [hunk 6]
```diff
@@ -1866,25 +1814,59 @@ PyDoc_STRVAR(module_doc,
...
         return -1;
     }
     return 0;
 }
 
+static int
+_jsonmodule_traverse(PyObject *module, visitproc visit, void *arg)
+{
+    _jsonmodulestate *state = (_jsonmodulestate *)PyModule_GetState(module);
+    if (state) {
```
Comment (vstinner) [inline Modules/_json.c]: state cannot be NULL since https://bugs.python.org/issue39824: remove the if. Same remark for clear.

File: Modules/_json.c [hunk 7]
```diff
@@ -1866,25 +1814,59 @@ PyDoc_STRVAR(module_doc,
...
+        Py_VISIT(get_json_state(module)->PyScannerType);
```
Comment (vstinner) [inline Modules/_json.c]: You just read the state, why getting it again? Just use state, no?

File: Modules/_json.c [hunk 8]
```diff
@@ -1866,25 +1814,59 @@ PyDoc_STRVAR(module_doc,
...
+    if (PyModule_AddObject(module, "make_encoder", get_json_state(module)->PyEncoderType) < 0) {
         return -1;
     }
     return 0;
 }
 
+static int
+_jsonmodule_traverse(PyObject *module, visitproc visit, void *arg)
+{
+    _jsonmodulestate *state = (_jsonmodulestate *)PyModule_GetState(module);
```
Comment (vstinner) [inline Modules/_json.c]: I suggest to use get_json_state() here: state cannot be NULL.

File: Modules/_json.c (hunk 3 above)
Comment (vstinner) [reply Modules/_json.c:643]: Oh. I missed the fact that PyScanner_Check() requires PyScannerType, but it's stored in the module state and there is no way to access the module state here :-( There would be two way to access the module state:

* Get it from Py_TYPE(self): sadly, currently, types don't store a reference to their module
* Get it with PyState_FindModule(). Sadly, this API doesn't work with modules using the multiphase initialization.

This problem is solved by the PEP 573 which has just been accepted, but not implemented yet.

I don't think that it's worth it to keep the code commented until the PEP 573 is implemented. These assertions were just for sanity checks, but it's unlikely that we get an object of a different type here. I mean, that it's not worth it to bother with these assertions.

File: Modules/_json.c [hunk 9]
```diff
@@ -1762,86 +1736,58 @@ encoder_listencode_list(PyEncoderObject *s, _PyAccu *acc,
 static void
 encoder_dealloc(PyObject *self)
 {
+    PyTypeObject *tp = Py_TYPE(self);
     /* bpo-31095: UnTrack is needed before calling any callbacks */
     PyObject_GC_UnTrack(self);
-    encoder_clear(self);
-    Py_TYPE(self)->tp_free(self);
+    encoder_clear((PyEncoderObject *)self);
+    tp->tp_free(self);
```
Comment (vstinner) [inline Modules/_json.c:1740]: The current code is just fine, no? I don't see the value of tp. It seems like it comes from a previous change that you reverted.

```suggestion
    Py_TYPE(self)->tp_free(self);
```

File: Modules/_json.c [hunk 10]
```diff
@@ -1209,47 +1209,24 @@ scanner_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
-    0,                    /* tp_base */
-    0,                    /* tp_dict */
-    0,                    /* tp_descr_get */
-    0,                    /* tp_descr_set */
-    0,                    /* tp_dictoffset */
-    0,                    /* tp_init */
-    0,/* PyType_GenericAlloc, */        /* tp_alloc */
-    scanner_new,          /* tp_new */
-    0,/* PyObject_GC_Del, */              /* tp_free */
+PyType_Spec PyScannerType_spec = {
```
Comment (vstinner) [inline Modules/_json.c]: nitpick: please put static on the same line, no need to write this on two lines.

File: Modules/_json.c [hunk 11]
```diff
@@ -1275,7 +1252,8 @@ encoder_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
         return NULL;
     }
 
-    s = (PyEncoderObject *)type->tp_alloc(type, 0);
+    allocfunc alloc_func = PyType_GetSlot(type, Py_tp_alloc);
+    s = (PyEncoderObject *)alloc_func(type, 0);
```
Comment (vstinner) [inline Modules/_json.c]: There is no need to use PyType_GetSlot(), just revert this change.

File: Modules/_json.c [hunk 12]
```diff
@@ -1866,25 +1812,54 @@ PyDoc_STRVAR(module_doc,
 static int
 _json_exec(PyObject *module)
 {
```
Comment (vstinner) [inline Modules/_json.c:1811]: I dislike calling get_json_state(module) multiple times. I would prefer to have a `_jsonmodulestate *state = get_json_state(module);` variable and reuse it.

File: Modules/_json.c [hunk 13]
```diff
@@ -1866,25 +1812,54 @@ PyDoc_STRVAR(module_doc,
...
-    if (PyType_Ready(&PyScannerType) < 0) {
+    PyObject *PyScannerType = PyType_FromSpec(&PyScannerType_spec);
```
Comment (vstinner) [inline Modules/_json.c]: I don't think that it's worth it to have a temporary variable: just writ into `state->PyScannerType` directly.

File: Modules/_json.c (hunk 9 above)
Comment (corona10) [reply Modules/_json.c:1740]: @vstinner 
`Py_DECREF(tp);`is needed if not test is leaked.
This is why I declared tp ;)

Comment (corona10) [reply Modules/_json.c:1740]:
```bash
test_json leaked [114, 114, 114] references, sum=342
test_json failed in 35.6 sec
```

**Comment by corona10** [comment]
@vstinner 
Thanks for the review. :)
I 've applied all of your comment except https://github.com/python/cpython/pull/19177#discussion_r398917087 :)

Please take a look

**Comment by vstinner** [comment]
Thanks @corona10, this change was interesting. I learnt a few things :-) I merged your PR, but I completed the commit message to elaborate on changes that you wrote.

**Comment by vstinner** [comment]
I mentioned the removed assertions in bpo-40137 that I just created: TODO list when PEP 573 "Module State Access from C Extension Methods" will be implemented.
